File: atcoder/abc339/d.py
from collections import deque


N = int(input())
arr = [input() for _ in range(N)]
base = [1, N, N**2, N**3]


def encode(v):
    return sum(x * y for x, y in zip(v, base))


def decode(v):
    state = []
    for _ in range(4):
        state.append(v % N)
        v //= N
    return state


rs = [0, 0]
cs = [0, 0]
i = 0
for r in range(N):
    for c in range(N):
        if arr[r][c] == "P":
            rs[i] = r
            cs[i] = c
            i += 1

que = deque()
dis = dict()

root = encode([rs[0], cs[0], rs[1], cs[1]])
que.append(root)
dis[root] = 0

while len(que) > 0:
    k = que.popleft()
    d = dis[k]
    (r0, c0, r1, c1) = decode(k)

    if (r0, c0) == (r1, c1):
        print(d)
        exit()

    for dr, dc in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
        rr0, cc0 = r0, c0
        rr1, cc1 = r1, c1

        nr, nc = r0 + dr, c0 + dc
        if 0 <= nr < N and 0 <= nc < N and arr[nr][nc] != "#":
            rr0, cc0 = nr, nc

        nr, nc = r1 + dr, c1 + dc
        if 0 <= nr < N and 0 <= nc < N and arr[nr][nc] != "#":
            rr1, cc1 = nr, nc

        key = encode([rr0, cc0, rr1, cc1])
        if key not in dis:
            dis[key] = d + 1
            que.append(key)

print(-1)

File: atcoder/abc339/d.rs
#![allow(unused)]

fn main() {
    let n = read::<usize>();
    let mut ban = std::collections::HashSet::new();
    let mut rs = vec![0; 2];
    let mut cs = vec![0; 2];
    let mut i = 0;
    for r in 0..n {
        let row = reads();
        for c in 0..n {
            if row[c] == '#' {
                ban.insert((r as i32, c as i32));
            } else if row[c] == 'P' {
                rs[i] = r as i32;
                cs[i] = c as i32;
                i += 1;
            }
        }
    }

    let mut dis = std::collections::HashMap::new();
    let mut que = std::collections::VecDeque::new();
    let root = ((rs[0], cs[0], rs[1], cs[1]));
    dis.insert(root, 0);
    que.push_back(root);

    while let Some((r0, c0, r1, c1)) = que.pop_front() {
        let d = dis[&(r0, c0, r1, c1)];

        if (r0, c0) == (r1, c1) {
            println!("{}", d);
            return;
        }

        for (dr, dc) in [(0, 1), (0, -1), (1, 0), (-1, 0)] {
            let (nr0, nc0) = (r0 + dr, c0 + dc);
            let (nr1, nc1) = (r1 + dr, c1 + dc);
            let ok1 = (nr0 >= 0 && nr0 < n as i32 && nc0 >= 0 && nc0 < n as i32);
            let ok2 = (nr1 >= 0 && nr1 < n as i32 && nc1 >= 0 && nc1 < n as i32);
            let ok3 = !ban.contains(&(nr0, nc0));
            let ok4 = !ban.contains(&(nr1, nc1));
            let mut rr0 = r0;
            let mut cc0 = c0;
            if ok1 && ok3 {
                (rr0, cc0) = (nr0, nc0);
            }
            let mut rr1 = r1;
            let mut cc1 = c1;
            if ok2 && ok4 {
                (rr1, cc1) = (nr1, nc1);
            }

            let key = (rr0, cc0, rr1, cc1);
            if !dis.contains_key(&key) {
                dis.insert(key, d + 1);
                que.push_back(key);
            }
        }
    }

    println!("-1");
}

fn read<T: std::str::FromStr>() -> T {
    let mut s = String::new();
    std::io::stdin().read_line(&mut s).ok();
    s.trim().parse().ok().unwrap()
}

fn readv<T: std::str::FromStr>() -> Vec<T> {
    read::<String>()
        .split_ascii_whitespace()
        .map(|t| t.parse().ok().unwrap())
        .collect()
}

fn reads() -> Vec<char> {
    read::<String>().chars().collect::<Vec<char>>()
}

fn mapv<T, S, F: Fn(&T) -> S>(arr: &Vec<T>, f: F) -> Vec<S> {
    arr.iter().map(f).collect()
}

fn join<T: ToString>(arr: &[T], sep: &str) -> String {
    arr.iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join(sep)
}
